app/javascript/lib/validation/chunked-line-reader.js

```diff
@@ -0,0 +1,101 @@
+import { DEFAULT_CHUNK_SIZE, readFileBytes } from './io'
+
+const newlineRegex = /\r?\n/
+
+/**
+ * reads lines from a file in a chunked way, so that the file does not have to be in memory all at once
+ * Basic usage:
+ * const chunker = new ChunkedLineReader(file)
+ * chunker.iterateLines((line, lineNum, isLastLine) => {
+ *   // do stuff
+ * }) {
+ * }
+ */
+export default class ChunkedLineReader {
+  /** create a new chunk reader */
+  constructor(file, chunkSize=DEFAULT_CHUNK_SIZE) {
+    this.file = file
+    this.nextByteToRead = 0
+    this.chunkSize = chunkSize
+    this.linesRead = 0
+    this.currentFragment = null // currentFragment stores the parts of lines that cross chunk boundaries
+    this.chunkLines = []
+    this.updateHasMoreChunks()
+    this.updateHasMoreLines()
+  }
+
+  /**
+   * iterates over the remaining lines in the file, calling func(line, lineNum, isLastLine) for each line
+   * (lineNum is 0-indexed).
+   * maxBytesPerLine can be set to avoid reading the entire file into memory in the event the file is missing
+   * proper newlines
+  */
+  async iterateLines(func, maxLines=Number.MAX_SAFE_INTEGER, maxBytesPerLine=1000*1000*1024) {
```

Comment:
Rather than defining the default value as an expression, I would assign this to a constant and reference it.

```suggestion
  const MAX_BYTES_PER_LINE = 1000*1000*1024
  async iterateLines(func, maxLines=Number.MAX_SAFE_INTEGER, maxBytesPerLine=MAX_BYTES_PER_LINE) {
```